h3py/h3core.pyx

```diff
@@ -0,0 +1,274 @@
+        res = resolution(h) - 1
+    # todo: actually, do we want to raise an error if there are no children, or just return an empty set?
+    u.check_res(res)
+
+    return h3c.h3ToParent(h, res)
+
+
+cpdef int distance(H3int h1, H3int h2) except -1:
+    """ compute the hex-distance between two hexagons
+    """
+    u.check_addr(h1)
+    u.check_addr(h2)
+
+    d = h3c.h3Distance(h1,h2)
+
+    return d
+
+cpdef H3int[:] disk(H3int h, int k):
+    """ Return cells at grid distance `<= k` from `h`.
+
+
+    """
+    u.check_addr(h)
+    if k < 0:
+        raise H3ValueError('Invalid ring size: {}'.format(k))
+
+    n = h3c.maxKringSize(k)
+
+    ptr = create_ptr(n) # todo: return a "smart" pointer that knows its length?
+    h3c.kRing(h, k, ptr)
+    mv = create_mv(ptr, n)
+
+    return mv
+
+cpdef H3int[:] ring(H3int h, int k):
+    """ Return cells at grid distance `== k` from `h`.
+
+    Collection is "hollow" for k >= 1.
+
+    """
+    u.check_addr(h)
+
+    n = 6*k if k > 0 else 1
+    ptr = create_ptr(n)
+
+    flag = h3c.hexRing(h, k, ptr)
+    mv = create_mv(ptr, n)
+
+    # todo: we can do this much more efficiently by using kRingDistances
+    if flag != 0:
+        raise H3ValueError("Couldn't run the fast version!")
+        # # fall back to `kRingDistances` and filter for appropriate distance. don't need to use sets
+        # n = h3c.maxKringSize(k)
+        # ptr = create_ptr(n)
+
+        # # hmmm. maybe use a cython array here instead
+        # cdef int[:n] distances
+        # distances[:] = 0
+
+
+        # h3c.kRingDistances(h, k, ptr, &distances[0])
+
+    return mv
+
+
+
+cpdef H3int[:] children(H3int h, res=None):
+    u.check_addr(h)
+
+    if res is None:
+        res = resolution(h) + 1
+    # todo: actually, do we want to raise an error if there are no children, or just return an empty set?
+    u.check_res(res)
+
+    n = h3c.maxH3ToChildrenSize(h, res)
+
+    ptr = create_ptr(n)
+    h3c.h3ToChildren(h, res, ptr)
+    mv = create_mv(ptr, n)
+
+    return mv
+
+
+
+cpdef H3int[:] compact(const H3int[:] hu):
+    for h in hu:
+        u.check_addr(h)
+
+    ptr = create_ptr(len(hu))
+    flag = h3c.compact(&hu[0], ptr, len(hu))
+    mv = create_mv(ptr, len(hu))
+
+    if flag != 0:
+        raise H3ValueError('Could not compact set of hexagons!')
+
+    return mv
+
+# todo: https://stackoverflow.com/questions/50684977/cython-exception-type-for-a-function-returning-a-typed-memoryview
+# apparently, memoryviews are python objects, so we don't need to do the except clause
+cpdef H3int[:] uncompact(const H3int[:] hc, int res):
+    for h in hc:
+        u.check_addr(h)
+
+    N = h3c.maxUncompactSize(&hc[0], len(hc), res)
+
+    ptr = create_ptr(N)
+    flag = h3c.uncompact(
+        &hc[0], len(hc),
+           ptr, N,
+        res
+    )
+    mv = create_mv(ptr, N)
+
+    if flag != 0:
+        raise H3ValueError('Could not uncompact set of hexagons!')
+
+    return mv
+
+
+
+# weird return type here
```

Comment:
Just use `int64_t` directly here rather than H3int? The type in the C library is signed (unlike H3int)

Reply:
fixed.